HIVE-27850: Major Compaction for Iceberg tables

difin opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Hive Iceberg Major Compaction

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
To add a major compaction feature to Hive Iceberg.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### Is the change a dependency upgrade?
<!--
If yes, please attach a file with output from mvn dependency:tree to validate a complete upgrade of dependency.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Manual testing on the local MiniHS2Cluster and developed a new query test.

ayushtkn commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+-- Mask neededVirtualColumns due to non-strict order
+--! qt:replace:/(\s+neededVirtualColumns:\s)(.*)/$1#Masked#/
+
+-- set hive.llap.io.enabled=true;
+-- set hive.vectorized.execution.enabled=true;
+-- set hive.optimize.shared.work.merge.ts.schema=true;
```
why are these commented?

difin replied:
Done

ayushtkn commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+    last_name STRING
+ )
+STORED BY ICEBERG STORED AS ORC 
+TBLPROPERTIES ('engine.hive.enabled'='true','format-version'='2');
```
why this is needed ```'engine.hive.enabled'='true'```, it should be by default true & no other test does that

difin replied:
Done

ayushtkn reviewed (commented):
Thanx @difin, dropped some comments, we should even assert for the output of ``Show Compactions`` & the compaction related tables to see everything is good over there.

zhangbutao commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -469,12 +473,29 @@ private Long getSnapshotId(Table table, String branchName) {
-      FilesForCommit results, Operation operation) {
+                           FilesForCommit results, Operation operation, boolean isCompaction) {
+
+    if (isCompaction) {
```
If i want to specify the output data file size using iceberg property `target-file-size-bytes`, how to do?
Can we refer to https://iceberg.apache.org/docs/latest/spark-procedures/#rewrite_data_files to allow users to specify property to rewrite data files?

ayushtkn replied:
@difin during iceberg table write, there must be some configuration which will be deciding the file size for normal data writes. So, if you set that conf, you should be able to achieve what is configured as part of ``target-file-size-bytes``?

So, the new files would be written of size mentioned as part of ``target-file-size-bytes``?

This could be 3 commits:
Alter table to set the write size if the size is specified?
Do the truncate
Do the insert?

simhadri-g commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorCompaction.java:
```diff
@@ -0,0 +1,60 @@
+    String tableName = context.getTable().getTableName();
+    LOG.debug("Executing compaction query for table {}", tableName);
+
+    String insertIntoTempIcebergTable = String.format("insert into %s select * from %s", tableName, tableName);
```
Small query, Since compaction may take some time to complete(for large tables), if there is any parallel writes to this table it will create a new snapshot right? How do we determine the correct snapshot to use(Do we lock table to prevent parallel writes)?

difin replied:
We don't lock the table for writes, but if there are parallel writes, only one of the writes will succeed to commit.

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -469,12 +473,29 @@ private Long getSnapshotId(Table table, String branchName) {
+      write.commit();
+
+      transaction.commitTransaction();
+      return;
```
Done

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorCompaction.java:
```diff
@@ -0,0 +1,60 @@
+    String tableName = context.getTable().getTableName();
+    LOG.debug("Executing compaction query for table {}", tableName);
+
+    String insertIntoTempIcebergTable = String.format("insert into %s select * from %s", tableName, tableName);
```
Done

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorCompaction.java:
```diff
@@ -0,0 +1,60 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IcebergMajorCompaction extends QueryCompactor  {
```
Done

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -437,6 +439,8 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
       deleteFiles.addAll(writeResults.deleteFiles());
       referencedDataFiles.addAll(writeResults.referencedDataFiles());
+
+      isCompaction = isCompaction || jobContext.getJobConf().getBoolean(HiveIcebergMetaHook.ICEBERG_COMPACTION, false);
```
Done

aturoczy commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -413,6 +415,7 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
 
     Table table = null;
     String branchName = null;
+    boolean isCompaction = false;
```
I'm bit problematic with the variable names. 
IsComapction in my mind does not express rightly the variable and the domain model.
I would suggest:

- isCompactionEnabled
- compactionEnabled
- comapctionStatus (Enum in this case maybe better)
- isCompactionActive
- allowCompaction 

I think those variable + getter setter would much more express the importance of this variable. cc: @ayushtkn @dengzhhu653 Please disagree if this is not the hive style.

dengzhhu653 replied:
I'm ok with the name, looks this variable tends to use internally only in this class(no set/get).

aturoczy commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java:
```diff
@@ -0,0 +1,60 @@
+  public boolean run(CompactorContext context) throws IOException, HiveException, InterruptedException {
+
+    String compactTableName = context.getTable().getTableName();
+    LOG.debug("Executing compaction query for table {}", compactTableName);
```
_Initiating compaction for the 'TableName' table._

I guess it bit expressive. (Also not sure, it is for table or on table :) )

difin replied:
Thanks, incorporated the suggestion with '... for the table ... '.

aturoczy reviewed (changes requested):
minor changes requested

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorCompaction.java:
```diff
@@ -0,0 +1,60 @@
+  private boolean runQuery(HiveConf conf, String query, String user) {
+
+    SessionState sessionState = DriverUtils.setUpSessionState(conf, user, true);
+    conf.setBoolean(HiveIcebergMetaHook.ICEBERG_COMPACTION, true);
```
Done

difin commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+Update ice_orc set last_name = 'ln4a' where first_name='fn4';
+Update ice_orc set last_name = 'ln5a' where first_name='fn5';
+Update ice_orc set last_name = 'ln6a' where first_name='fn6';
+Update ice_orc set last_name = 'ln7a' where first_name='fn7';
```
Done.

difin commented on itests/qtest-iceberg/src/test/java/org/apache/hadoop/hive/cli/TestIcebergLlapLocalCompactorCliDriver.java:
```diff
@@ -0,0 +1,83 @@
+/*
```
Because existing IcebergCli doesn't start a compaction worker.
When a compaction is run on a regular driver without compaction worker the tests hangs forever.
Hive ACID also has separate driver for compaction q-tests.

difin commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorFactory.java:
```diff
@@ -108,6 +110,23 @@ public CompactorPipeline getCompactorPipeline(Table table, HiveConf configuratio
+        case MAJOR:
+
+          try {
+            Class<? extends QueryCompactor> inputFormatClass = (Class<? extends QueryCompactor>)
```
Done.. didn't notice that after copy pasting..

difin commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+Update ice_orc set last_name = 'ln6a' where first_name='fn6';
+Update ice_orc set last_name = 'ln7a' where first_name='fn7';
+
+alter table ice_orc COMPACT 'major' and wait;
```
done

difin commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+ )
+STORED BY ICEBERG STORED AS ORC 
+TBLPROPERTIES ('engine.hive.enabled'='true','format-version'='2');
+
```
Added

difin commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction.q:
```diff
@@ -0,0 +1,33 @@
+-- set hive.llap.io.enabled=true;
+-- set hive.vectorized.execution.enabled=true;
+-- set hive.optimize.shared.work.merge.ts.schema=true;
+
```
Done

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -437,6 +440,10 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
       deleteFiles.addAll(writeResults.deleteFiles());
       referencedDataFiles.addAll(writeResults.referencedDataFiles());
+
+      isCompaction = isCompaction || Optional.ofNullable(SessionState.get())
```
why it's in a loop?

difin replied:
Fixed. 
It's was a leftover from previous variations, previously the compaction flag was taken from job conf and there is a list of them in OutputTable.

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLUtils.java:
```diff
@@ -235,6 +235,10 @@ public static boolean isIcebergTable(Table table) {
   }
 
+  public static boolean isIcebergTable(org.apache.hadoop.hive.metastore.api.Table table) {
+    return table.getParameters().get(HiveMetaHook.TABLE_TYPE).equals(HiveMetaHook.ICEBERG);
```
To avoid NPE:
```suggestion
    return HiveMetaHook.ICEBERG.equals(table.getParameters().get(HiveMetaHook.TABLE_TYPE));
```

difin replied:
Thanks! Done.

zhangbutao commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java:
```diff
@@ -0,0 +1,60 @@
+    String compactTableName = context.getTable().getTableName();
+    LOG.debug("Executing compaction query for table {}", compactTableName);
+
+    String insertIntoTempIcebergTable = String.format("insert into %s select * from %s",
```
Why not `insret overwrite`？

difin replied:
IOW causes problems if an Iceberg table had partition evolution and it is not supported ([HIVE-26133](https://issues.apache.org/jira/browse/HIVE-26133)). The workaround in such case is to do truncate before IOW.
And this is what is done in this PR.
In case of compaction request, the commitWrite method was extended to do these 2 operations in a single transaction: 

1. Start a new transaction.
2. Delete all existing files in the manifest (equivalent to truncate table operation).
3. Add new files to the manifest.
4. Commit transaction.

deniskuzZ replied:
please use IOW, but handle it properly in commitOverwrite

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLUtils.java:
```diff
@@ -235,6 +235,10 @@ public static boolean isIcebergTable(Table table) {
             table.getStorageHandler().getType() == StorageHandlerTypes.ICEBERG;
   }
 
+  public static boolean isIcebergTable(org.apache.hadoop.hive.metastore.api.Table table) {
```
please use method from MetaStoreUtils

difin replied:
Done

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -469,12 +474,34 @@ private Long getSnapshotId(Table table, String branchName) {
-      FilesForCommit results, Operation operation) {
+                           FilesForCommit results, Operation operation, boolean isCompaction) {
+
+    if (isCompaction) {
```
could you please extract this logic into existing commitOverwrite method and in case of MAJOR compaction use enum REWRITE_ALL_PARTITIONS.
In future we might map new DDL operation `OPTIMIZE TABLE REWRITE DATA` to that logic in order to fix issues with partition evolution + rewrite small files.

deniskuzZ replied:
````
enum RewriteData {DEFAULT, ALL_PARTITIONS}
````
pass this instead of isCompaction()

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactOperation.java:
```diff
@@ -52,7 +60,7 @@ public AlterTableCompactOperation(DDLOperationContext context, AlterTableCompact
   @Override public int execute() throws Exception {
     Table table = context.getDb().getTable(desc.getTableName());
-    if (!AcidUtils.isTransactionalTable(table)) {
+    if (!AcidUtils.isTransactionalTable(table) && !DDLUtils.isIcebergTable(table)) {
```
use AcidUtils.isNonNativeAcidTable(table) instead

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorFactory.java:
```diff
@@ -108,6 +111,24 @@ public CompactorPipeline getCompactorPipeline(Table table, HiveConf configuratio
           throw new HiveException(
               compactionInfo.type.name() + " compaction is not supported on insert only tables.");
       }
+    } else if (DDLUtils.isIcebergTable(table)) {
```
use method from MetaStoreUtils.isIcebergTable

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java:
```diff
@@ -0,0 +1,60 @@
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive.compaction;
```
should be in `org.apache.hadoop.hive.ql.txn.compactor` package of ql module

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -553,7 +573,11 @@ private void commitOverwrite(Table table, String branchName, long startTime, Fil
     }
-
+    if (rewritePolicy != RewritePolicy.DEFAULT) {
+      LOG.info("Completed compaction for table {}, number of files committed: {}",
```
That's not generic - 'compaction'. That msg should be logged on the QueryCompactor side

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/Context.java:
```diff
@@ -250,6 +252,24 @@ public String toString() {
+    public static RewritePolicy fromString(String rewritePolicy) {
+      
+      if (rewritePolicy == null) {
+        throw new IllegalArgumentException("Invalid RewritePolicy: null");
```
Preconditions.checkArgument(null != rewritePolicy, "Invalid rewrite policy: null")

difin replied:
done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/Context.java:
```diff
@@ -90,6 +91,7 @@ public class Context {
   String originalTracker = null;
   private CompilationOpContext opContext;
   private final Map<String, ContentSummary> pathToCS = new ConcurrentHashMap<String, ContentSummary>();
+  public static final String REWRITE_POLICY = "hive.rewrite.policy";
```
please move it to `InputFormatConfig.class`: REWRITE_POLICY="hive.rewrite.data.policy";

difin replied:
done

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -437,6 +445,9 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
       dataFiles.addAll(writeResults.dataFiles());
       deleteFiles.addAll(writeResults.deleteFiles());
       referencedDataFiles.addAll(writeResults.referencedDataFiles());
+      rewritePolicy = isCompaction && rewritePolicy == RewritePolicy.DEFAULT ?
```
no need to check if isCompaction
````
RewritePolicy rewritePolicy = null;
for (JobContext jobContext : outputTable.jobContexts) {
 branchName = conf.get(InputFormatConfig.OUTPUT_TABLE_SNAPSHOT_REF);
rewritePolicy = RewritePolicy.fromString(conf.get("REWRITE_POLICY", RewritePolicy.DEFAULT.name()));
....
}
````

difin replied:
done

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java:
```diff
@@ -0,0 +1,63 @@
+
+    SessionState sessionState = DriverUtils.setUpSessionState(conf, user, true);
+    sessionState.setCompaction(true);
+    conf.set(Context.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
```
HiveConf.setVar(conf, HiveConf.ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());

difin replied:
done

deniskuzZ reviewed (commented):
LGTM, some minor comments

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -531,16 +538,25 @@ private void commitWrite(Table table, String branchName, Long snapshotId, long s
         overwrite.toBranch(HiveUtils.getTableSnapshotRef(branchName));
       }
       overwrite.commit();
+      transaction.commitTransaction();
```
there is no conflict validation and branch support, see commitWrite. However, we can do this in the next PRs.

difin replied:
Ackowledged, will handle it in the next commits.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:
```diff
@@ -68,6 +68,22 @@ abstract class QueryCompactor implements Compactor {
   protected void commitCompaction(String dest, String tmpTableName, HiveConf conf,
       ValidWriteIdList actualWriteIds, long compactorTxnId) throws IOException, HiveException {}
 
+  protected SessionState setupQueryCompaction(HiveConf conf, CompactionInfo compactionInfo, Map<String, String> tblProperties) {
```
setupQueryCompactionSession

difin replied:
Done

difin commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -437,6 +441,9 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
       dataFiles.addAll(writeResults.dataFiles());
       deleteFiles.addAll(writeResults.deleteFiles());
       referencedDataFiles.addAll(writeResults.referencedDataFiles());
+      rewritePolicy = rewritePolicy == RewritePolicy.DEFAULT ?
```
Because this assignment is happening inside the loop which iterates over the list of job contexts.
I am assigning the rewrite policy in case it is still default (unset) because if there are >1 job context, which one of them will contain the rewrite policy config.

deniskuzZ replied:
in case of IOW, should would be just 1 job context

difin replied:
Done.

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:
```diff
@@ -453,7 +456,13 @@ private void commitTable(FileIO io, ExecutorService executor, OutputTable output
+
+      RewritePolicy rewritePolicy = RewritePolicy.fromString(outputTable.jobContexts.stream()
+          .findAny()
+          .map(x -> x.getJobConf().get(ConfVars.REWRITE_POLICY.varname, RewritePolicy.DEFAULT.name()))
```
why do you define defaults twice? you already have `orElse`

difin replied:
Fixed

deniskuzZ commented:
@difin, are we using ACID thresholds to schedule the compaction, Worker#isEnoughToCompact? Also, after the successful compaction, do we mark the entry as "ready for cleaning" or move it into the "COMPLETED_COMPACTIONS", CompactionTxnHandler#markCleaned?

deniskuzZ commented on iceberg/iceberg-handler/src/test/queries/positive/iceberg_major_compaction_partitioned.q:
```diff
@@ -0,0 +1,38 @@
+explain alter table ice_orc COMPACT 'major' and wait;
+alter table ice_orc COMPACT 'major' and wait;
+
+select * from ice_orc;
```
what do we test here? should we print the `snapshot-summary` to check there are no delete files?

difin replied:
Added printing of `snapshot-summary` to the tests. Also updated `HiveIcebergStorageHandler` to allow IOW after partition evolution in case of Iceberg Compaction.

difin commented:
> @difin, are we using ACID thresholds to schedule the compaction, Worker#isEnoughToCompact? Not even sure if that lists the proper data dir. Also, after the successful compaction, do we mark the entry as "ready for cleaning" or move it into the "COMPLETED_COMPACTIONS", CompactionTxnHandler#markCleaned?

I implemented a strategy design pattern in the main routine of the compaction worker to have different behavior of the worker for different tables types. Also, I added `show compactions ` command to the new q-tests to test that iceberg compactions get to completed stage now.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/ACIDCompactionExecutor.java:
```diff
@@ -0,0 +1,267 @@
+  private final boolean collectMrStats;
+  private final Worker.CompactionTxn compactionTxn;
+
+  public ACIDCompactionExecutor(Worker worker, CompactionInfo ci, Worker.CompactionTxn compactionTxn,
```
why do you need `Worker` here? move needed methods here or in `CompactionExecutor`

deniskuzZ replied:
should only get msc client and conf

difin replied:
Worker is here because compaction executor code (Iceberg and ACID) that was copied from Worker uses many methods from Worker and CompactorThread (parent of Worker) classes which are also used in another placed. below is the list of these methods:

```
CompactorThread.resolveStorageDescriptor()
CompactorThread.checkInterrupt()
CompactorThread.runJobAsSelf()
CompactorThread.resolvePartition()
CompactorThread.isMinorCompactionSupported
Worker.markFailed()

```
So we can't just move these methods to CompactionExecutor, we would have to copy the code and it would be duplication. Please advise how you would like me to proceed here.

deniskuzZ replied:
can't we move those to some common interface for both CompactorThread and CompactionExecutor

deniskuzZ replied:
maybe rename existing TaskHandler to CleanupTaskHandler and move common methods into new TaskHandler?
cc @SourabhBadhya

difin replied:
Hi @deniskuzZ,
I did the refactoring, moved the common methods into a separate common class and got rid of the Worker in CompactionExecutors. I moved all the common compaction methods into CompactionUtil class.
I think the best location for all the common compaction methods is not in the TaskHandler, but in the CompactionUtil because CompactionUtil already contained several duplicated methods which were also present in CompactorThread and TaskHandler is just another client of CompactionUtil which uses many methods from there, like CompactorThread also does.
I refactored TaskHandler too and made it use resolvePartition method from CompactionUtil were I moved it from CompactorThread.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,124 @@
+  private final boolean collectMrStats;
+  private final Worker.CompactionTxn compactionTxn;
+
+  public IcebergCompactionExecutor(Worker worker, CompactionInfo ci, Worker.CompactionTxn compactionTxn, 
```
there is no compactionTxn &  collectMrStats  for Iceberg

difin replied:
boolean genericStats = conf.getBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS);
    boolean mrStats = conf.getBoolVar(HiveConf.ConfVars.HIVE_MR_COMPACTOR_GATHER_STATS);

difin replied:
genericStats and mrStats values are decided based on HiveConf values in the Worker, why is mrStats not relevant for Iceberg?

```
    boolean genericStats = conf.getBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS);
    boolean mrStats = conf.getBoolVar(HiveConf.ConfVars.HIVE_MR_COMPACTOR_GATHER_STATS);
```

deniskuzZ replied:
`mr` means - MapReduce, that mode won't be supported for Iceberg tables

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionExecutor.java:
```diff
@@ -0,0 +1,26 @@
+
+public interface CompactionExecutor {
+  
+  Boolean executeCompaction() throws InterruptedException, TException, IOException;
```
boolean compact(CompactionInfo ci)

difin replied:
done.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -337,151 +341,16 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
-        LOG.warn(ci.errorMessage + " Compaction info: {}", ci);
-        msc.markRefused(CompactionInfo.compactionInfoToStruct(ci));
-        return false;
+      if (MetaStoreUtils.isIcebergTable(table.getParameters())) {
```
use factory to create a proper strategy
cc @veghlaci05 on generic redesign

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -499,67 +368,24 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
       }
     }
 
-    if (computeStats) {
```
that should stay here

difin replied:
I moved `computeStats` it to CompactionExecutor and returned the logic to update stats at the end of `findNextCompactionAndExecute` as following:

```
    if (Optional.ofNullable(compactionExecutor).map(CompactionExecutor::isComputeStats).orElse(false)) {
      statsUpdater.gatherStats(ci, conf, runJobAsSelf(ci.runAs) ? ci.runAs : table.getOwner(),
          CompactorUtil.getCompactorJobQueueName(conf, ci, table), msc);
    }
```

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -499,67 +368,24 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
   }
 
-  private void markFailed(CompactionInfo ci, String errorMessage) {
+  protected void markFailed(CompactionInfo ci, String errorMessage) {
```
should stay private, executor should return exception

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -337,151 +341,16 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
-        computeStats = false;
-
-        if (runJobAsSelf(ci.runAs)) {
-          cleanupResultDirs(sd, tblValidWriteIds, ctype, dir);
```
should stay

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -83,7 +70,7 @@ public class Worker extends RemoteCompactorThread implements MetaStoreThread {
 
   private String workerName;
-  private final CompactorFactory compactorFactory;
+  protected final CompactorFactory compactorFactory;
```
is that use at all?

difin replied:
Not used in Worker. Moved to CompactionExecutor.

deniskuzZ commented:
isRebalanceCompaction is only ACID specific

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionExecutorFactory.java:
```diff
@@ -0,0 +1,39 @@
+
+public class CompactionExecutorFactory {
+  
+  public static CompactionExecutor getInstance(Table table, Worker worker, Worker.CompactionTxn compactionTxn, 
```
```` 
 private static final CompactionExecutorFactory INSTANCE = new CompactionExecutorFactory();

  public static CompactionExecutorFactory getInstance() {
    return INSTANCE;
  }
````

deniskuzZ replied:
the current method should be 'createExecutor(Table table)'

difin replied:
Done.

I named getInstance based on other code examples. 
For example, in TaskHandlerFactory class.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/ACIDCompactionExecutor.java:
```diff
@@ -0,0 +1,294 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ACIDCompactionExecutor extends CompactionExecutor {
```
AcidCompactionExecutor

difin replied:
Done

difin commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -260,8 +187,9 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
-    boolean computeStats = false;
     Table table = null;
+    boolean compactionResult = false;
+    CompactionExecutor compactionExecutor = null;
```
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,102 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IcebergCompactionExecutor extends CompactionExecutor {
```
should it be in iceberg module?

difin replied:
Yes, done.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,102 @@
+    // Find the appropriate storage descriptor
+    final StorageDescriptor sd =  CompactorUtil.resolveStorageDescriptor(table);
+
+    if (isTableSorted(sd, ci)) {
```
remove this check

difin replied:
Done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,102 @@
+
+    CompactorUtil.checkInterrupt(CLASS_NAME);
+    
+    msc.updateCompactorState(CompactionInfo.compactionInfoToStruct(ci), DEFAULT_TXN_ID);
```
we don't need this

difin replied:
removed

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,102 @@
+    msc.updateCompactorState(CompactionInfo.compactionInfoToStruct(ci), DEFAULT_TXN_ID);
+
+    // Don't start compaction or cleaning if not necessary
+    if (isDynPartAbort(table, ci)) {
```
not applicable for Iceberg, please remove

difin replied:
done

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/IcebergCompactionExecutor.java:
```diff
@@ -0,0 +1,102 @@
+    CompactorUtil.checkInterrupt(CLASS_NAME);
+
+    try {
+      failCompactionIfSetForTest();
```
please remove this, that's a hack in old code

difin replied:
done

deniskuzZ reviewed (approved):
LGTM +1, but i would rename these 3 from `xxxExecutor`
````
CompactionService
AcidCompactionService
IcebergCompactionService
````

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/RemoteCompactorUtil.java:
```diff
@@ -0,0 +1,62 @@
+  static final private String CLASS_NAME = RemoteCompactorUtil.class.getName();
+  protected static final Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
+
+  public static Table resolveTable(HiveConf conf, IMetaStoreClient msc, CompactionInfo ci) throws MetaException {
```
how is that different from CompactorUtil.resolveTable?

difin commented:
> LGTM +1, but i would rename these 3 from `xxxExecutor`
> 
> ```
> CompactionService
> AcidCompactionService
> IcebergCompactionService
> ```

I've renamed the classes.
Thanks for approval!
